### ISSUE TITLE
Add hasNotNullValueSatisfying()

### ISSUE THREAD

**Issue by wimdeblauwe** [assertj/assertj-core#2392]
#### Summary

When `Optional` is used, you can use `assertThat(optional).hasValueSatisfying( value -> ...)`. It would be convenient to also have something similar for nullable types.

Sometimes, I have assertions like this in my tests:

```java
assertThat(thing.getSomething()).isNotNull();
assertThat(thing.getSomething().getValue()).isEqualTo(...);
```

#### Example

It would be nice to be able to do the above in 1 statement like this:

```java
assertThat(thing.getSomething().getValue()).hasNotNullValueSatisfying( value -> assertThat(value).isEqualTo(...));
```

Note: not sure if this is a problem, but this would mean that people could use this on an `Optional` as well, where the `value` in the lambda would be the optional itself. That is one of the reasons why I would propose to use a slightly different method name compared to `hasValueSatisfying`.

**Comment by scordio** [comment]
Have you tried [`get()`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractOptionalAssert.html#get()) (or [`get(InstanceOfAssertFactory)`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractOptionalAssert.html#get(org.assertj.core.api.InstanceOfAssertFactory)) for type-narrowed assertions)?

It performs an implicit null check and allows to write something like:

```java
assertThat(thing.getSomething())
    .get()
    .isEqualTo(...);
```

**Comment by wimdeblauwe** [comment]
But then I can't do `something.getValue()` inside of the `isEqualsTo`, right?

**Comment by scordio** [comment]
What is `getValue()` exactly? Are you referring to [`Optional#get()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#get--)?

**Comment by wimdeblauwe** [comment]
No, `getValue()` is just a random example of a getter on something that can be `null`.

Maybe it is clearer like this:

Suppose you have this:

```
assertThat(thing.getSomething()).isEqualTo("something");
```

If `thing` can be `null`, you need to guard against it like this:

```
assertThat(thing).isNotNull();
assertThat(thing.getSomething()).isEqualTo("something");
```

I would love to be able to do this:

```
assertThat(thing)
    .hasNotNullValueSatisfying( thing -> assertThat(thing.getSomething()).isEqualTo("Something"));
```

(Writing it like this, I wonder if `isNotNullValueSatisfying()` would be a better method name)

**Comment by scordio** [comment]
Thanks, now it's clear to me. Currently, the easiest way I see is:

```java
assertThat(thing).isNotNull()
                 .extracting(ThingType::getSomething)
                 .isEqualTo("something");
```
Unfortunately, `extracting` does not perform a null check in advance, so 

```java
assertThat(thing).extracting(ThingType::getSomething).isEqualTo("something");
```

would fail with a `NullPointerException` if `thing` is null.

What if we would enhance the behavior of `extracting`?

**Comment by BenTels** [comment]
> What if we would enhance the behavior of `extracting`?

I would rethink that -- you'd be changing the semantics of an existing method. Right now `extracting` assumes that actual is not null or the extractor set by the developer can handle nulls. If anybody has used that fact, you'll be silently breaking their tests.

You could add an `extractingNullSafe` or something. But I actually think the meaning of @wimdeblauwe 's proposal is clearer.

Either way, I would improve the documentation of the existing `extracting` method. :smile:

**Comment by scordio** [comment]
What I meant is that `extracting` should throw a proper assertion error with the corresponding message if the value is null, instead of just failing with a `NullPointerException`. I doubt that someone is relying on its current behavior in case of a null value, most likely there is a `isNotNull` call in front of it.

**Comment by BenTels** [comment]
That's also an option, I guess (although I would still leave it alone).

I still think @wimdeblauwe 's idea is better.

**Comment by scordio** [comment]
The point is to reduce the options to do the same thing, which is already a bit of a flaw with AssertJ.

Currently, I don't see many benefits of introducing a new API for just a preliminary null check, while there is already an existing method that should have done it implicitly.

But let's reevaluate this if other use cases arise.

**Comment by wimdeblauwe** [comment]
With `extracting`, there is no way to nest the calls like with my proposal:

```
assertThat(thing)
    .hasNotNullValueSatisfying( thing -> {
            assertThat(thing.getSomething())
                    .hasNotNullValueSatisfying( something -> {
                            assertThat(something.getSomeOtherThing()).isEqualTo("SomeOtherThing");
```

(That said, I still think `extracting` should not fail with a `NullPointerException`, but I see that as a separate issue)

About:
> The point is to reduce the options to do the same thing, which is already a bit of a flaw with AssertJ.

I don't see it as a flow that we have `isPresent`, `isNotPresent`, `isEmpty` and `isNotEmpty` for asserting `Optional`. I like having the option to use the method that makes the test the most readable given my current use case.

**Comment by scordio** [comment]
> With `extracting`, there is no way to nest the calls like with my proposal:

Why not? Chaining is already a feature and only the null check is currently missing.

Assuming implicit null-check, you could write:

```java
assertThat(thing)
  .extracting(ThingType::getSomething) // fails with AssertionError if thing is null
  .extracting(SomethingType::getSomeOtherThing) // fails with AssertionError if getSomething() result is null
  .isEqualTo("SomeOtherThing")
```

or, sacrificing type safety:


```java
assertThat(thing)
  .extracting("something.someOtherThing") // fails with AssertionError if any intermediate value is null
  .isEqualTo("SomeOtherThing")
```

**Comment by joel-costigliola** [comment]
I prefer @scordio suggestion with extracting as it is IMO more elegant than:
```
assertThat(thing)
    .hasNotNullValueSatisfying( thing -> {
            assertThat(thing.getSomething())
                    .hasNotNullValueSatisfying( something -> {
                            assertThat(something.getSomeOtherThing()).isEqualTo("SomeOtherThing");
```

**Comment by wimdeblauwe** [comment]
Unless you want to have multiple assertions at the same "level":

```
assertThat(thing)
    .hasNotNullValueSatisfying( thing -> {
            assertThat(thing.getSomethingString()).isEqualTo("foo");
            assertThat(thing.getSomethingList()).hasSize(4);
            assertThat(thing.getSomething())
                    .hasNotNullValueSatisfying( something -> {
                            assertThat(something.getSomeOtherThing()).isEqualTo("SomeOtherThing");
```

**Comment by scordio** [comment]
I think that we are heading to style details. The previous example could be written as:

```java
assertThat(thing).isNotNull();
assertThat(thing.getSomethingString()).isEqualTo("foo");
assertThat(thing.getSomethingList()).hasSize(4);
assertThat(thing.getSomething()).isNotNull();                            assertThat(thing.getSomething().getSomeOtherThing()).isEqualTo("SomeOtherThing");
```

I don't think we can objectively say that the nested solution is more readable than the flat one. Both are equally verbose and it really depends on personal taste.

For this reason, I still don't see it as a valuable addition.
